map/chart_generator.cpp

```diff
@@ -0,0 +1,215 @@
+}
+
+void GenerateChartByPoints(size_t width, size_t height, vector<m2::PointD> const & geometry,
+                           bool day, vector<uint8_t> & frameBuffer)
+{
+  frameBuffer.clear();
+  if (width == 0 || height == 0 || geometry.empty())
+    return;
+
+  agg::rgba8 const kBackgroundColor = agg::rgba8(255, 255, 255, 0);
+  agg::rgba8 const kLineColor =
+      day ? agg::rgba8(30, 150, 240, 255) : agg::rgba8(255, 230, 140, 255);
+  agg::rgba8 const kCurveColor = day ? agg::rgba8(30, 150, 240, 20) : agg::rgba8(255, 230, 140, 20);
+  double constexpr kLineWidthPxl = 2.0;
+  uint32_t constexpr kBPP = 4;
+
+  using TBlender = BlendAdaptor<agg::rgba8, agg::order_rgba>;
+  using TPixelFormat = agg::pixfmt_custom_blend_rgba<TBlender, agg::rendering_buffer>;
+  using TBaseRenderer = agg::renderer_base<TPixelFormat>;
+  using TPrimitivesRenderer = agg::renderer_primitives<TBaseRenderer>;
+  using TSolidRenderer = agg::renderer_scanline_aa_solid<TBaseRenderer>;
+  using TPath = agg::poly_container_adaptor<vector<m2::PointD>>;
+  using TStroke = agg::conv_stroke<TPath>;
+
+  agg::rendering_buffer renderBuffer;
+  TPixelFormat pixelFormat(renderBuffer, agg::comp_op_src_over);
+  TBaseRenderer m_baseRenderer(pixelFormat);
+
+  frameBuffer.assign(width * kBPP * height, 0);
+  renderBuffer.attach(&frameBuffer[0], static_cast<unsigned int>(width),
+                      static_cast<unsigned int>(height), static_cast<int>(width * kBPP));
+  m_baseRenderer.reset_clipping(true);
+  unsigned op = pixelFormat.comp_op();
+  pixelFormat.comp_op(agg::comp_op_src);
+  m_baseRenderer.clear(kBackgroundColor);
+  pixelFormat.comp_op(op);
+
+  agg::rasterizer_scanline_aa<> rasterizer;
+  rasterizer.clip_box(0, 0, width, height);
+
+  // Polygon under chart line.
+  agg::path_storage underChartGeometryPath;
+  underChartGeometryPath.move_to(geometry.front().x, static_cast<double>(height));
+  for (auto const & p : geometry)
+    underChartGeometryPath.line_to(p.x, p.y);
+  underChartGeometryPath.line_to(geometry.back().x, static_cast<double>(height));
+  underChartGeometryPath.close_polygon();
+
+  agg::conv_curve<agg::path_storage> curve(underChartGeometryPath);
+  rasterizer.add_path(curve);
+  agg::scanline32_p8 scanline;
+  agg::render_scanlines_aa_solid(rasterizer, scanline, m_baseRenderer, kCurveColor);
+
+  // Chart line.
+  TPath path_adaptor(geometry, false);
+  TStroke stroke(path_adaptor);
+  stroke.width(kLineWidthPxl);
+  stroke.line_cap(agg::round_cap);
+  stroke.line_join(agg::round_join);
+
+  rasterizer.add_path(stroke);
+  agg::render_scanlines_aa_solid(rasterizer, scanline, m_baseRenderer, kLineColor);
+}
+
+void GenerateChart(size_t width, size_t height, deque<double> const & distanceDataM,
+                   feature::TAltitudes const & altitudeDataM, bool day,
+                   vector<uint8_t> & frameBuffer)
+{
+  frameBuffer.clear();
+  if (altitudeDataM.empty() || distanceDataM.size() != altitudeDataM.size())
```

Comment:
GenerateChart looks like a Public API while functions above looks like a private implementation helpers. May be move them into anonymous namespace then?